🚀 2단계 - 리팩터링(메뉴)

jivebreaddev opened the pull request:
안녕하세요 웅래님!

Application 레이어 까지 진행해서 수정량이 너무 많아졌네요;;
application 레이어 코드가 줄어드는것을 참을수가 없었네요!! 
아직도 더 줄일수 있는 방법이 있으면 말씀해주세요!

다음 미션에는 domain만 진행하겠습니다;; 너무나 많은 코드 리뷰드리게 되서 죄송합니다!!

## 질문들

1. 간접 참조로 빼고 어떤 값들을 수동 업데이트 해야되는지 확인했습니다. 서로 간접 참조로 관계를 생성하였습니다.
- menu product가 product에 간접 참조하게 되고 price 값이 수동으로 update되어야하는 점입니다. (이벤트를 받거나, 수동으로 table update 필요)


2. factory 랑 validator, domain service??를 적용할 수 있는 곳들이 있을까요?

3. ProductPrice와 MenuPrice 계속 동일한 로직 반복되는것같은데 공통 객체로 common 도메인으로 빼서 사용하시나요?

4. application 레벨에서 event 구현 해도 괜찮을까요??

6. ID도 VO로 생성해야되나요??


![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/9ea68903-81dc-433b-97ea-5492b85b6588)

jivebreaddev committed:
feat(Product): Product VO인 ProductPrice, ProductName 추가
- ProductName의 규약 검증을 위해서 ProfanityValidator를 생성시 추가
- DIP를 위해서 Interface를 domain에 추가

jivebreaddev committed:
test(Product): Product -> ProductTest로 수정하다가 잘못 refactor 된부분들 수정

jivebreaddev committed:
test(Product): Product, ProductName, ProductPrice 도메인 레이어 테스트 생성
- 규약에 적힌 규칙들에 대해 테스트

jivebreaddev committed:
test(Product): 안쓰는 add 제거

jivebreaddev committed:
feature(ProductService): domain 패키지에서 product를 dto로 변경하고 리팩토링한 product로 변경하였습니다.

jivebreaddev committed:
feature(Menu): Repository 및 의존성 변경 && entity fields 생성했습니다.

Jivebread committed:
feat(DIP) domain 패키지로 interface를 이전하고 JPA REPOSITORY들은 INFRA로 이동한다.

Jivebread committed:
feat(tobe-domain) domain 패키지로 VO 객체들을 생성하고 2개의 aggregate들을 생성하였습니다. 그리고 menu 와 menu group으로 aggregate을 나눴습니다.
1. AGGREGATE - MENU
  a. menu product
  b. menu name
  c. menu price
2. AGGREGATE - MENU GROUP
  a. menu group
  b. menu group name

마지막으로, 서로 간접 참조로 관게를 생성하였습니다.
- 여기서 중요한 점은, menu product가 product에 간접 참조하게 되고 price 값이 수동으로 update되어야하는 점입니다. (이벤트를 받거나, 수동으로 table update 필요)

Jivebread committed:
feat(tobe-domain) application 레이어로 dto 이전

Jivebread committed:
feat(tobe-domain) 1. 주석제거

Jivebread committed:
feat(tobe-domain): Fixtures 와 ProductFixture 통합으로 Fixtures으로 이전

Jivebread committed:
test(ProductTest): static 필드 값을 명시적 값으로 변경

Jivebread committed:
test(Repository, Validator): 가짜객체들 별도의 패키지(infra)로 분리

Jivebread committed:
feature(README): 도메인 모델링 사항 반영

Jivebread committed:
Merge branch 'step1' into step2

# Conflicts:
#	src/test/java/kitchenpos/menus/application/MenuServiceTest.java

Jivebread committed:
feature(README): 도메인 모델링 사항 반영

Jivebread committed:
feature(ProductService): productService 에서 menu에 대한 로직 제거

Jivebread committed:
feature(MenuGroup & MenuGroupTest): MenuGroup과 VO 생성 및 TEST 생성했습니다.

Jivebread committed:
feature(MenuGroup & MenuGroupTest): 필요없는 IMPORT 제거

Jivebread committed:
feature(MenuGroup & MenuGroupTest): 모델링에서 메뉴 카테고리의 영어이름 MENU GROUP으로 변경하였습니다.

Jivebread committed:
feature(MenuGroup): Equals & Hash 생성

Jivebread committed:
feature(MenuName): Equals & Hash 생성

Jivebread committed:
test(MenuGroupTest): Menugroup 도메인 테스트 생성하였습니다.

Jivebread committed:
FEATURE(README): 모델링 수정 -> 난잡한 언어 간소화

Jivebread committed:
FEATURE(Menu): 원래의 menu를 request로 수정하고 Menu도메인 객체 생성

Jivebread committed:
FEATURE(Menu): 원래의 menu를 request로 수정하고 Menu도메인 객체 생성

Jivebread committed:
FEATURE(Menu): Menu VO 생성

Jivebread committed:
FEATURE(Menu): 원래 MenuProducts를 MenuProductsRequest로 변경

Jivebread committed:
FEATURE(Menu): 원래 MenuGroup을 MenuGroupRequest로 변경

Jivebread committed:
FEATURE(Menu): 원래 MenuGroup을 MenuGroupRequest로 변경

Jivebread committed:
FEATURE(Menu): Menu 도메인 로직 구현

Jivebread committed:
FEATURE(Menu): Menu Application 로직 구현

Jivebread committed:
Test(MenuPriceTest): MenuPrice에 대한 테스트 구현

Jivebread committed:
Feature(Order): Order에서 Menu -> MenuRequest 에 따른 사이드이펙트 수정

Jivebread committed:
test(Test): Test 실패에 따른 수정들 진행했습니다.
- menu service test
- menu test
진행 필요

Jivebread committed:
chores(merge): merge conflict 제거
- 머지 컨플릭트 제거

Jivebread committed:
test(JPAMenuRepository): JPA 오류 제거
- ENTITY 연관관계 맵핑 오류, JPQL 에러 수정했습니다.

Jivebread committed:
feature(Menu): 메뉴의 vo들을 가져오는 로직중 long 과 bigdecimal 두개다 지원할 수 있도록 변경하였습니다.

Jivebread committed:
test(MenuTest): 메뉴들에 대한 테스트를 실행 될수있도록 수정하였습니다.

Jivebread committed:
feature(menu): EQUALS & HASHCODE 생성

Jivebread committed:
test(MenuTest): 메뉴들에 대한 테스트를 실행 될수있도록 수정하였습니다.

wlroh commented on README.md:
```diff
@@ -107,8 +107,8 @@ docker compose -p kitchenpos up -d
 |------------|--------------------|------------------------------------------|
-| 메뉴 카테고리    | Menu Category      | 메뉴들의 집합 (ex: '추천메뉴', '신메뉴', '2명을 위한 메뉴') |
-| 메뉴 카테고리 이름 | Menu Category Name | 손님들에게 보여지는 메뉴들의 집합의 명칭                   |
+| 메뉴 카테고리    | Menu Group         | 메뉴들의 집합 (ex: '추천메뉴', '신메뉴', '2명을 위한 메뉴') |
+| 메뉴 카테고리 이름 | Menu Group Name | 손님들에게 보여지는 메뉴들의 집합의 명칭                   |
```
꾸준한 용어사전 업데이트 👍 👍 💯

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/Menu.java:
```diff
@@ -40,37 +33,21 @@ public class Menu {
-  )
-  private MenuGroup menuGroup;
-
+  @Column(name = "menu_group_id", columnDefinition = "binary(16)", nullable = false)
+  private UUID menuGroupId;
```
메뉴 그룹 애그리거트와 메뉴 애그리거트로 구현해주셨네요 👍 
다른 애그리거트 간접참조 👍 👍 💯

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuProducts.java:
```diff
@@ -0,0 +1,18 @@
+  private List<MenuProduct> values = new ArrayList<>();
+
+
+}
```
일급컬렉션 👍 👍

jivebreaddev replied:
저 17 라인의 빈공간들이 신경이 쓰이네요 하하하;;

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/Menu.java:
```diff
@@ -1,35 +1,19 @@
 @Entity
 public class Menu {
 
+  public static final int ZERO = 0;
```
public 으로 접근제어자를 설정해주신 의도가 있을까요? 🤔

jivebreaddev replied:
접근제어자 확인하겠습니다!

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuProducts.java:
```diff
@@ -4,15 +4,44 @@
+            .reduce(BigDecimal.ZERO, (a, b) -> a.add(b));
+
+  }
+  @EventListener
```
`@EventListener` 를 적어주신 의도가 있을까요? 🤔

jivebreaddev replied:
여담이지만 
'상품'에서 가격을 변경하였을때 
'메뉴'에서도 가격이 변경되는것을 감지하고 'display상태'가  변경되는것을 이벤트로 구현하면 되지않을까 하며 마커만 해뒀습니다!

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuProduct.java:
```diff
@@ -32,7 +32,7 @@ private MenuProduct(UUID productId, MenuProductPrice price, Integer quantity) {
-    if (quantity > ZERO) {
+    if (quantity <= ZERO) {
       throw new IllegalArgumentException("`메뉴`에 등록된 `상품 개수`는 1개 이상이어야 한다.");
     }
```
테스트코드의 중요성을 다시 한번 느끼네요 👍

jivebreaddev replied:
간단한 도메인 로직을 잘못쓸수도 있었고 테스트가 잡아줬네요! 헛헛;;

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuPrice.java:
```diff
@@ -0,0 +1,66 @@
+  private MenuPrice(BigDecimal price) {
+    validate(price);
+    this.price = price;
+  }
```
주생성자 부생성자 방식으로 적용해보면 어떨까요?

jivebreaddev replied:
`  private Price(final Long price) {
    this(BigDecimal.valueOf(Objects.requireNonNull(price)));
  }`

이렇게 변경했었을것같네요!


현재는 값객체가 하나로 통일하면서 factory 메서드로 다 변환 시켰습니다!

![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/005a604b-4b7f-4997-aa2d-057c8aaa76d5)

jivebreaddev replied:
`주생성자 부생성자 생성으로 객체를 생성하기위해 하나의 로직만 남김으로써 통일된 객체 상태를 만든다!`

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuProducts.java:
```diff
@@ -0,0 +1,65 @@
+            .forEach(menuProduct -> menuProduct.changePrice(price));
+  }
+
+  public void add(MenuProduct menuProduct){
+    this.products.add(menuProduct);
+  }
+
+  public boolean findByProductId(UUID productId){
+    return this.products.stream().anyMatch(product -> product.getId().equals(productId));
+  }
+
+  public boolean containsZeroProducts(){
+    return this.products.isEmpty();
+  }
```
애그리거트간의 통신을 하기 위해서는 루트 어그리거트를 통해서만 통신이 가능해야한다고 생각하는데요!
휴먼에러를 막기 위해 접근제어자를 package-private 으로 설정해보는 것은 어떤가요?

jivebreaddev replied:
최소한의 접근제어를 확인 못했네요 

적용하겠습니다!!!

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuProductPrice.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Objects;
+
+@Embeddable
+public class MenuProductPrice {
```
MenuPrice 와 MenuProductPrice 가 결국 둘다 Price 이고 역할이 동일하다고 생각하는데 하나의 값객체를 사용해보는 것은 어떨까요?

jivebreaddev replied:
해당 정책은 하나의 값객체를 사용해보겠습니다!

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuName.java:
```diff
@@ -0,0 +1,54 @@
+package kitchenpos.menus.domain.tobe.menu;
+
+import jakarta.persistence.Embeddable;
+import kitchenpos.products.domain.tobe.ProfanityValidator;
```
현재 비속어 검증을 product 도메인에 있는 것을 사용하고 있는데요 
비속어 검증을 메뉴컨텍스트에도 구성해보거나, Name 값객체를 공통으로 분리해보면 어떨까요?

jivebreaddev replied:
중복의 경우 공통으로 분리하거나 비속어 검증을 컨텍스트 마다 생성함

Name 의 정책이 달라지는 경우 낭패가 생길 가능성이 높아 보여서 validator 만 추가해보겠습니다!

wlroh commented:
> 간접 참조로 빼고 어떤 값들을 수동 업데이트 해야되는지 확인했습니다. 서로 간접 참조로 관계를 생성하였습니다.
menu product가 product에 간접 참조하게 되고 price 값이 수동으로 update되어야하는 점입니다. (이벤트를 받거나, 수동으로 table update 필요)

menuProduct 가 product 를 간접참조 👍 👍  혹시 어떤 점이 궁금하셨을까요? 🤔 

> factory 랑 validator, domain service??를 적용할 수 있는 곳들이 있을까요?

메뉴생성에서 현재 메뉴그룹에 속해있어야 한다까지를 도메인 지식으로 보았고 유효한 메뉴그룹인지에 대해서는 비즈니스적으로 유의미하지 않다고 생각하셔서 application 계층에서 exception 을 발생하고 있는데요!
만약 메뉴를 생성할 때 메뉴 그룹도 유효한 메뉴그룹인지 판단하는 것이 비즈니스적으로 유의미하다고 생각되시면 factory 를 통해 menu 를 생성해보는 것도 괜찮다고 생각해요!

> ProductPrice와 MenuPrice 계속 동일한 로직 반복되는것같은데 공통 객체로 common 도메인으로 빼서 사용하시나요?

값객체의 경우 common 으로 빼서 사용하기도 하는데요. common 은 양날의 검이여서 PR 리뷰와 같이 개발문화에 따라 common 으로 분리해서 관리할지 각 컨텍스트마다 관리할지 고민해볼 것 같아요 😄 

> application 레벨에서 event 구현 해도 괜찮을까요??

event 객체를 말씀하시는 것일까요? 발행을 의미하는 것일까요? 저는 보통 이벤트 객체의 경우 여러 컨텍스트에서 이벤트 객체를 의존해야하는 경우가 많아서  common 패키지에 구현하는 것을 선호해요

> ID도 VO로 생성해야되나요??

ID 도 Value Object 로 구성하는 것은 강제할 필요는 없지만, 명시적으로 ID 라는 Value Object 로 구성해주는 것도 괜찮다고 생각해요

wlroh reviewed (changes requested):
시창님 안녕하세요 😄 
미션 고생하셨습니다 👍 
코멘트 작성한 것이 있는데 확인부탁드려요!

jivebreaddev commented:
1. 비속어 검증 aggregate 마다 인터페이스 생성
2. 가격 값객체로 공통사용
3. 접근 제어자들 확인
4. 메뉴 가격 생성시 생성자 부생성자 활용
5. 추후에 팩터리 메서드로 애플리케이션 로직 마저 내리기

질문들에서 해결된점

1. 질문이 아닌데 질문으로 넣어뒀네요

2.`메뉴생성에서 현재 메뉴그룹에 속해있어야 한다까지를 도메인 지식으로 보았고 유효한 메뉴그룹인지에 대해서는 비즈니스적으로 유의미하지 않다고 생각하셔서 application 계층에서 exception 을 발생하고 있는데요!`
-> 애플리케이션 계층에서 두개의 에그리게이트에 대한 예외가 발생하는 경우도 팩터리 메서드를 통한 도메인 로직 이전이 가능하군요!



3. `event 객체를 말씀하시는 것일까요? 발행을 의미하는 것일까요? 저는 보통 이벤트 객체의 경우 여러 컨텍스트에서 이벤트 객체를 의존해야하는 경우가 많아서 common 패키지에 구현하는 것을 선호해요`

의존성이 도메인간에 생기는 것보다는 common 으로 향하는 것을 선호하시군요! 질문이 구체적이지 않았는데 찰떡같이 구현방법을 설명해주셔서 감사합니다!!

Jivebread committed:
feature(Menu): Zero 에 대한 접근제어자 제거

Jivebread committed:
feature(ProfanityValidator): 비속어 검증 로직에 대한 인터페이스를 공통으로 이전했습니다.

Jivebread committed:
feature(Price): 가격(Price)을 공통 값객체로 변경하였습니다.

Jivebread committed:
feature(Price): final 키워드 수정

Jivebread committed:
feature(ProductPriceChange): ProductPriceChangeEvent 구현했습니다.

product 가 변경되면 ProductID를 동일하게가진 메뉴에 대해서 값이 변경되고

Jivebread committed:
test(ProductPriceChange): ProductPriceChangeEvent에 대한 검증을 진행했습니다.

jivebreaddev commented:
![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/985240fa-be07-4c43-9d51-7cbd0b3c5f08)

Jivebread committed:
feature(MenuDomainService): MenuDomainService 로 여러개의 Aggregate에 대한 로직을 생성하였습니다.

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/Menu.java:
```diff
@@ -0,0 +1,129 @@
+      throw new IllegalArgumentException("`메뉴 상품 가격`의 총액보다 `메뉴 가격`이 클 수 없다.");
+    }
+
+    if (menuProducts.sum().compareTo(BigDecimal.ZERO) < ZERO) {
+      throw new IllegalArgumentException("`메뉴`의 `메뉴 가격`은 양수이어야한다.");
+    }
```
요 검증에 대해서는 이미 Price 값객체가 양수임을 보장하기 때문에 없어도 괜찮다고 생각해요!

jivebreaddev replied:
넵 코드가 불필요하게 있는 로직들은 제거가 되어야 겠네요!

wlroh commented on src/main/java/kitchenpos/common/domain/ProfanityValidator.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.products.domain.tobe;
+package kitchenpos.common.domain;
```
이번 미션이 domain 계층에 대해 리팩터링하는 것이여서 아시는 내용인데 수정을 안하셨을 것 같은데요!
비속어에 대한 검증을 현재 product 컨텍스트 내 infra 에 구현이 되어 있어요!
혹시 그런 의도시면 괜찮지만, 아니였다면 고민해보시는 것도 좋을 것 같아요! 이부분은 꼭 반영하실필요는 없어요!

jivebreaddev replied:
1. 문제:
/product/infra/DefaultProfanityValidator 로 구현되어있는 곳으로 의존성 방향이 menu -> product가 생김

2. 해결:
- 인터페이스는 common/domain에 두고 
- /menu/infra/MenuProfanityValidator를 구현해서 의존성 방향이 menu -> menu로 생기게 변경

wlroh commented on src/main/java/kitchenpos/menus/application/MenuService.java:
```diff
@@ -1,144 +1,80 @@
         return menu;
     }
 
+    @TransactionalEventListener
+    @Async
+    public void changeMenuProdutPrice(final ProductPriceChangeEvent productPriceChangeEvent) {
+        final List<Menu> menus = menuRepository.findAllByProductId(productPriceChangeEvent.getProductId());
+
+        for (final Menu menu : menus) {
+            menu.getMenuProducts().changeMenuProductPrice(productPriceChangeEvent.getProductId(), productPriceChangeEvent.getPrice());
+
+            if (menu.getMenuPrice().compareTo(menu.getMenuProducts().sum()) > 0) {
+                menu.hide();
+            }
+        }
+    }
```
현재 상품 가격이 변경되었을 때 메뉴가 미노출될 수 가 있는데 트랙잭션이 끝나있어서 제대로 반영이 안될 것으로 보여요 😢 
어떤 의도로 리스너 설정해주신 것일까요?

wlroh replied:
> menu.getMenuProducts().changeMenuProductPrice(productPriceChangeEvent.getProductId(), productPriceChangeEvent.getPrice());

현재 루트어그리거트를 통해 수행되기보다는 메뉴상품에게 가격을 변경하라고 요청하고 있는데요! ACID 를 지켜주기 위해 루트어그리거트를 이용해보면 어떨까요?

wlroh replied:
또한, 메뉴 가격이 변경됨에 따라 메뉴가 미노출되는 것을 현재 application 서비스에서 제어해주고 있는데요!
가격이 변경됨에 따라 메뉴의 가격 정책에 따라 미노출되는 흐름이 하나의 중요한 비즈니스 로직이라고 생각해요
메뉴에게 가격변경이 이뤄지면 메뉴가 주도적으로 진행하도록 수정해보는 것은 어떤가요?

jivebreaddev replied:
1. package-private을 만들면서 자연스럽게 ACID를 지키게끔 변경해보겠습니다.

2. Application에서 Domain 으로 비즈니스 로직 이전

3. 상품 가격 변경일때, 트랜잭션 처리 확인

jivebreaddev replied:
`현재 상품 가격이 변경되었을 때 메뉴가 미노출될 수 가 있는데 트랙잭션이 끝나있어서 제대로 반영이 안될 것으로 보여요 😢
어떤 의도로 리스너 설정해주신 것일까요?`

![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/119fff11-0c16-4648-bf5e-be959806a6a7)
1. 트랜잭션 처리는 다음과 같이 변경하였습니다.

- TransacationalEventListener가 발동되는 시점에 entity가 변경되어도 수정이 반영되지 않는 이슈

a. 부모의 트랜잭션이후에 새로운 트랜잭션을 열어서 entity에 대한 update를 반영하는 방향으로 `REQUIRESNEW`을 주었습니다.

jivebreaddev replied:
비동기 이벤트의 경우에도 테스트가 못잡아주는군요;; 

굉장히 불안한 구현방법이네요!

jivebreaddev replied:
`또한, 메뉴 가격이 변경됨에 따라 메뉴가 미노출되는 것을 현재 application 서비스에서 제어해주고 있는데요!
가격이 변경됨에 따라 메뉴의 가격 정책에 따라 미노출되는 흐름이 하나의 중요한 비즈니스 로직이라고 생각해요
메뉴에게 가격변경이 이뤄지면 메뉴가 주도적으로 진행하도록 수정해보는 것은 어떤가요?`

![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/5b3148b2-4461-4c3f-bf82-9a21712a3f0c)


다음과 같이 변경하였습니다.

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuDomainService.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.UUID;
+
+@Component
+public class MenuDomainService {
```
저는 도메인서비스의 경우 ~~DomainService 보다는 더 명확하게 네이밍하는 것을 선호하는데 이번에 유의미한 네이밍으로 변경해보는 것은 어떤가요?

jivebreaddev replied:
네이밍 변경하겠습니다!

wlroh commented on src/main/java/kitchenpos/menus/domain/tobe/menu/MenuDomainService.java:
```diff
@@ -0,0 +1,72 @@
+    this.profanityValidator = profanityValidator;
+  }
+
+  public Menu createMenu(UUID menuGroupId, List<MenuProductRequest> menuProducts, boolean displayed, String name, BigDecimal price){
+
```
현재 Dto 가 도메인 계층까지 사용되면서 도메인 계층에서 외부에 대한 의존성이 생기고 있다고 보여져요. 😢 
도메인 계층을 견고하게 지키기 위해 고민해보시면 좋을 것 같아요
추가로, Product 를 메뉴에서 의존하는 것에 대해서도 고민해보시면 좋을 것 같아요

jivebreaddev replied:
도메인을 지키기 위한 의존성 체크
1. domain -> application (DTO)
2. Menu ->  Product (??) 확인 필요

wlroh replied:
메뉴컨텍스트에서 상품에 대해 관심사는 삼품의 ID와 가격이라고 생각해요!
그래서 interface 를 이용해 필요한 정보만 응답받도록 수정하시고 메뉴 컨텍스트에 구현체에서 상품 컨텍스트에 대한 의존을 가져가는 것 방향으로 가게되면 추후 상품이 변경되었을 때 혹은 MSA로 분리되더라도 infra 계층만 Http 통신으로 수정하고 메뉴컨텍스트의 도메인까지 영향을 안줄수 있다고 생각해요!

모놀리식서버의 장점을 활용하기 위해 상품컨텍스트의 의존성을 가져가는 것이 좋다고 생각하는데 확장성을 고려해서 interface와 값객체를 이용해  도메인계층 내 의존성을 끊어보는 것을 말씀드리고 싶었어요!

wlroh reviewed (changes requested):
시창님 안녕하세요 😄 
피드백 잘 반영해주셨습니다 👍 
코멘트 추가한 것이 있는데 확인부탁드려요!

jivebreaddev commented:
## 요구사항

1. 불필요한 검증 삭제 (price 관련 값객체)
2. ProfanityValidator로 인한 product/infra 로의 의존성들 확인 필요
3. product 값 변경으로 인한 이벤트에서 트랜잭션 처리 확인 필요 + 비즈니스 로직을 domain으로 내리기
4. 도메인 서비스 보다는 유의미한 factory 같은 명명으로 변경 필요
5. 도메인을 지키기 위한 의존성 체크
- domain -> application (DTO)
- Menu -> Product (??) 확인 필요
6. ACID 를 지키기위한 접근제어자 수정!!

Jivebread committed:
feature(MenuFactory): MenuDomainService -> MenuFactory로 이름을 변경하였습니다. + MenuService에 대한 의존성 주입 제거

Jivebread committed:
feature(ProfanityValidator): ProfanityValidator에 대한 의존도가 DefaultProfanityValidator(PRODUCT)로 향하는 것을 common/infra로 변경하였습니다.

Jivebread committed:
feature(Menu): 불필요한 검증로직을 제거했습니다. -> price에서 해당 값객체에 대한 검증이 있음

Jivebread committed:
feature(Menu): domain -> application (DTO)에대한 의존도를 제거하였습니다. (Product에 대한 의존도 제거하였습니다.)

Jivebread committed:
feature(MenuService): 메뉴가 주도적으로 가격 정책에 따라 전시상태를 미노출로 변경합니다.

jivebreaddev commented:
![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/899fb108-dc3f-496d-b50d-bb30489f790a)

Jivebread committed:
feature(MenuService): ACID 를 지키기 위해 루트에그리거트를 사용한 로직으로 변경하였습니다.

wlroh commented:
시창님 안녕하세요 😄 
질문주신 코멘트에 대해 답글 남겼어요! 
https://github.com/next-step/ddd-tactical-design/pull/304#discussion_r1632214926

추가로, 현재 메뉴상품가격에 변경됨에 따라 메뉴 노출이 변경되는 것이 모델링에는 반영이 안되어있어보여요!
기존 상품컨텍스트에 있던 내용이 구현하면서 모델링이 변경됨에 따라 수정된 것으로보여지는데 이런 현상은 매주 자연스럽다고 생각하지만,
모델링에 대해 업데이트도 꾸준히 부탁드릴게요!

모델링과 구현코드간의 싱크가 안맞으면 결국 도메인주도설계를 도입하기 전의 모습에서 발생했던 문제들이 계속 재발할 수 있다고 생각해요!

wlroh reviewed (changes requested):
시창님 안녕하세요 😄 
피드백 잘 반영해주셨습니다 👍 
고민있던 부분에 대해 코멘트 추가했는데 확인부탁드려요!

Jivebread committed:
feature(ProductClient): ProductClient 를 생성해서 MenuService에서 Product에 대한 의존도를 제거하였습니다.

Jivebread committed:
feature(ProductClient): MenuService 에서 ProductRepository를 제거하고 domain으로 검증로직을 내렸습니다.

Jivebread committed:
test(MenuService): 의존도 변경에 따른 test 수정을 진행하였습니다.
- MenuService에서 ProductRepository 제거
- MenuTest에서는 productClient를 주입하기위해서 repository에 상태를 생성해서 주입하고 진행하였습니다.

Jivebread committed:
test(modeling): 도메인 모델링에 메뉴 가격 관련 정책 수정하였습니다.

jivebreaddev commented:
![image](https://github.com/next-step/ddd-tactical-design/assets/89639413/52a63d2d-f613-40f3-a200-2088492bb2af)

`메뉴컨텍스트에서 상품에 대해 관심사는 삼품의 ID와 가격이라고 생각해요!
그래서 interface 를 이용해 필요한 정보만 응답받도록 수정하시고 메뉴 컨텍스트에 구현체에서 상품 컨텍스트에 대한 의존을 가져가는 것 방향으로 가게되면 추후 상품이 변경되었을 때 혹은 MSA로 분리되더라도 infra 계층만 Http 통신으로 수정하고 메뉴컨텍스트의 도메인까지 영향을 안줄수 있다고 생각해요!`

1. ProductClient를 인터페이스로 주입받아서 도메인 로직으로 내렸습니다.

`추가로, 현재 메뉴상품가격에 변경됨에 따라 메뉴 노출이 변경되는 것이 모델링에는 반영이 안되어있어보여요!
기존 상품컨텍스트에 있던 내용이 구현하면서 모델링이 변경됨에 따라 수정된 것으로보여지는데 이런 현상은 매주 자연스럽다고 생각하지만,
모델링에 대해 업데이트도 꾸준히 부탁드릴게요!

모델링과 구현코드간의 싱크가 안맞으면 결국 도메인주도설계를 도입하기 전의 모습에서 발생했던 문제들이 계속 재발할 수 있다고 생각해요!`

2. 모델링에 대한 수정도 진행하였습니다.



슬랙으로 많은 질문들에 답변을 정성껏 주셔서 감사드립니다!!

wlroh reviewed (approved):
시창님 안녕하세요 😄 
피드백 잘 반영해주셨습니다 👍 
그럼 마지막 단계도 화이팅입니다 🔥
